src/hotspot/share/code/vtableStubs.cpp

```diff
@@ -255,6 +255,19 @@ inline uint VtableStubs::hash(bool is_vtable_stub, int vtable_index){
 }
 
 
+inline uint VtableStubs::unsafe_hash(address entry_point) {
+  // The entrypoint may or may not be a VtableStub. Generate a hash as if it was.
+  address vtable_stub_addr = entry_point - VtableStub::entry_offset();
+  assert(CodeCache::contains(vtable_stub_addr), "assumed to always be the case");
```

Comment:
Why not `VtableStubs::contains()` ?

Reply:
This is not trying to ascertain if `vtable_stub_addr` is in a `VtableStub`. It may not be, because we do not yet have a megamorphic call. This whole patch is to handle this case. However, whatever our call destination was it must have been in the CodeCache. And not only that, we must maintain the invariant that 16 bytes before our call destination is also in the code cache, or we would not be able to dereference that memory.

I am not sure exactly how we ensure this. But this patch does not change the behaviour, we still look at exactly the same memory as before, just in a more C++ friendly way.